### PR TITLE
Java 1.21.3 Support

### LINKED ISSUE #929: Support Minecraft PC 1.21.3
A new Minecraft Java Edition version is available (as of 2024-10-23T12:28:15+00:00), version **1.21.3**
## Official Changelog
* https://feedback.minecraft.net/hc/en-us/sections/360001186971-Release-Changelogs
## Protocol Details
(I will close this issue automatically if "1.21.3" is added to data/pc/common/versions.json on "master")
<table>
  <tr><td><b>Name</b></td><td>1.21.3</td>
  <tr><td><b>Protocol ID</b></td><td>768</td>
  <tr><td><b>Release Date</b></td><td>2024-10-23T12:28:15+00:00</td>
  <tr><td><b>Release Type</b></td><td>release</td>
  <tr><td><b>Data Version</b></td><td>4082</td>
  <tr><td><b>Java Version</b></td><td>21</td>
</table>
<hr/>
🤖 I am a bot, I check for updates every 2 hours without a trigger. You can close this issue to prevent any further updates.

### PR DESCRIPTION
**Pull request by GroobleDierne**
Fixes #929, #928 

TODO:

- [x] Protocol
    - [x] packet_recipe_book_add
    - [x] packet_craft_recipe_response should be similar to previous one
    - [x] Fix Slot & SlotComponent types
    - [x] packet_declare_recipes
- [x] entities.json (waiting on minecraft-data-generator)
- [x] tints.json is likely incorrect

### CODE REVIEW

**Comment by rom1504** [comment]
Can you point missing files to 1.21.1 for now

**Comment by rom1504** [comment]
Let's open an nmp and mineflayer PRs using this

**Comment by rom1504** [comment]
nmp and mineflayer PRs

https://github.com/PrismarineJS/node-minecraft-protocol/pull/1347
https://github.com/PrismarineJS/mineflayer/pull/3489

fail due to wrong protocol as expected

**Comment by rom1504** [comment]
see https://github.com/extremeheat/extracted_minecraft_data/pulls as a source of info for the protocol

**Comment by rom1504** [comment]
https://github.com/extremeheat/extracted_minecraft_data/pull/13.patch

**Comment by GroobleDierne** [comment]
> see https://github.com/extremeheat/extracted_minecraft_data/pulls as a source of info for the protocol

I'm using yarn because some files failed to decompile in extremheat's repository.
Anyway, I'll stop here for tonight.

**Comment by rom1504** [comment]
https://github.com/PrismarineJS/node-minecraft-protocol/actions/runs/11543601911/job/32356083688 now passing but still some packet errors

**Comment by GroobleDierne** [comment]
> https://github.com/PrismarineJS/node-minecraft-protocol/actions/runs/11543601911/job/32356083688 now passing but still some packet errors

I still need to implement one or two server-bound packets and the `Slot` type is messed up which causes quite a lot of packets to throw errors sometimes. I'm also stuck with the `unlock_recipes` packet, @extremeheat could you take a look at it?

**Comment by rom1504** [comment]
https://github.dev/extremeheat/extracted_minecraft_data/pull/15 can help

File: data/pc/latest/proto.yml [hunk 1]
```diff
@@ -551,13 +536,12 @@
...
 
-   BlockSet:
+   IdSet:
       type: varint
       name: type ?
          if 0: string
          default: void
-      # TODO: This will require a custom type ; we need type - 1 length of blockIds
-      blockIds: type ?
+      identifier: type ?
```
Comment (extremeheat) [inline data/pc/latest/proto.yml:544]: Why rename of the field? Any field renames should be done to 1.21.1 also

File: data/pc/latest/proto.yml [hunk 2]
```diff
@@ -1542,7 +1531,40 @@
          if custom_report_details: packet_common_custom_report_details
          if server_links: packet_common_server_links
 ^play.toClient.types:
-   # World state information for spawn and respawn packets
+   packet_set_player_inventory:
```
Comment (extremeheat) [inline data/pc/latest/proto.yml:1534]: Please make sure to add up to date `# MC: ClassName` anchors to the top of all packet defs, this makes it easy to jump between the vanilla implementation and the protodef defs

**Comment by extremeheat** [comment]
I did some work on the protocol, still some work to do:

* Recipe changes (recipe_book_add/remove/etc.) not yet handled
* SlotComponent type needs fixes, likely back to 1.20.5
* Backport some fixes to 1.20.5
* Added new `registryEntryHolder` and `registryEntryHolderSet` types to correspond to the Holder and HolderSet types inside minecraft ([ref](https://github.com/extremeheat/extracted_minecraft_data/blob/client1.21.2/client/net/minecraft/network/codec/ByteBufCodecs.java#L817)) to replace usage of the `arrayWithLengthOffset` type
  * holderSet work by writing a varint to the stream, n, and 0 means "read one [string] entry" and non-zero n, means read an array of 1+n entries of data from the stream
  * holder work by writing a varint to the stream, n, and 0 means "read one [varint] entry" set to n+1 and non-zero means read an array of n entries of data from the stream
  * usage example (not yet implemented) to write a varint, n, and then either a varint or EntityMetadataWolfVariant:
 ``` yaml
  if wolf_variant: ["registryEntryHolder", {
       "base": { name: "variantId", type: "varint" },
       "otherwise": { name: "variantData", type: "EntityMetadataWolfVariant" }
  }]
  ```
* using the `bitflags` type [implemented in bedrock-protocol](https://github.com/PrismarineJS/bedrock-protocol/blob/fa730786d5f8c6c066fe221d92c99f4331b97dee/src/datatypes/compiler-minecraft.js#L123) to make the flags easier than tracking individual bits
```yaml
["bitflags", {
      "type": "u32",
      "flags": ["x", "y", "z", "yaw", "pitch", "dx", "dy", "dz", "yawDelta"]
   }]
```

File: data/pc/latest/proto.yml [hunk 3]
```diff
@@ -1229,16 +1208,6 @@
       details: []varint
          key: string
          value: string
-   # MC: ClientboundResourcePackPopPacket
-   packet_common_remove_resource_pack:
```
Comment (GroobleDierne) [inline data/pc/latest/proto.yml]: Why removing them from common?

File: data/pc/latest/proto.yml [hunk 4]
```diff
@@ -3689,8 +3096,12 @@
       actionId: varint
       jumpBoost: varint
    # MC: ServerboundPlayerInputPacket
-   packet_player_inputs:
-      inputsBitset: u8
+   packet_steer_vehicle:
```
Comment (GroobleDierne) [inline data/pc/latest/proto.yml:3099]: This name doesn't make sense anymore since this packet is sent even if the player isn't in a vehicle

**Comment by MrBruz** [comment]
@GroobleDierne For 'registryDataIsMandatory', 'version' needs to be changed to 'versions'

**Comment by GroobleDierne** [comment]
> @GroobleDierne For 'registryDataIsMandatory', 'version' needs to be changed to 'versions'

Thx

**Comment by GroobleDierne** [comment]
Now only recipes related packets seem to be missing for protocol support

**Comment by GroobleDierne** [comment]
I'll try debugging them later today

**Comment by rom1504** [comment]
Hey, please remove registries.json file from this PR and put it in its own PR. I think we need a lot more work on it to include in mc data

**Comment by rom1504** [comment]
Also let's review https://github.com/GroobleDierne/minecraft-data/pull/2

**Comment by GroobleDierne** [comment]
> Hey, please remove registries.json file from this PR and put it in its own PR. I think we need a lot more work on it to include in mc data

I realised that if the client and server share the same minecraft:core pack version (communicated through SelectKnowPacks), the server only needs to send every key from every registries. So my current plan is to implement this in p-registry, as for most registries, mc-data already has the data to generate the keys. In the long term, we must add some fields to the existing schemas and create new ones too.

**Comment by GroobleDierne** [comment]
After a looooooot of debugging, I figured out the bug was just in front of my eyes in the registryHolderSet decoding code and not in the protocol definition :face_holding_back_tears:

File: data/pc/latest/proto.yml [hunk 5]
```diff
@@ -116,17 +193,23 @@
...
          if custom_name: anonymousNbt
          if item_name: anonymousNbt
+         if item_model: string
          if lore: anonOptionalNbt[]varint
-         if rarity: varint
+         if rarity: varint =>
+         - common
+         - uncommon
+         - rare
+         - epic
```
Comment (extremeheat) [inline data/pc/latest/proto.yml:212]: We need to apply changes like this back to 1.20.5 also as to not make implementation in mineflayer messy

**Comment by Cleboost** [comment]
When do you think it will be released?

**Comment by TheJupiterDev** [comment]
What all needs done before this is done ?

**Comment by rom1504** [comment]
Lgtm I think all comments from extremeheat were fixed as well

Let's merge and iterate on further PRs

**Pull request merged**